### ISSUE TITLE
Angular 17.1, ag-grid 31.0.2 gridReady event not being triggered by await fixture.whenStable()

### ISSUE THREAD

**Issue by n19htz** [ag-grid/ag-grid#7463]
As the topic says, gridReady event not triggered anymore by sequence fixture.detectChanges(), await fixture.whenStable(). This basically breaking all existing tests with those scenarios.

**Comment by StephenCooper** [comment]
Did this break when upgrading from 31.0.1 -> 31.0.2?

**Comment by n19htz** [comment]
> Did this break when upgrading from 31.0.1 -> 31.0.2?

yes

reproduction: https://stackblitz.com/edit/stackblitz-starters-39ppuw?file=src%2Fapp-root.component.spec.ts

run in console:

```shell
npm run test
```

**Comment by n19htz** [comment]
app-root.component.ts

```typescript
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { Component, DestroyRef, inject } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { AgGridAngular } from 'ag-grid-angular';
import { ColDef, GridApi, GridReadyEvent } from 'ag-grid-community';
import { catchError, finalize, of, take } from 'rxjs';

interface IRow {
  id: number;
  name: string;
  email: string;
}

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [AgGridAngular, HttpClientModule],
  providers: [HttpClient],
  template: `
  <div class="content">
    <ag-grid-angular
      style="width: 100%; height: 350px;"
      class="ag-theme-material"
      [columnDefs]="colDefs"
      (gridReady)="onGridReady($event)"
    >
    </ag-grid-angular>
  </div>
`,
})
export class AppRootComponent {
  private readonly _httpClient = inject(HttpClient);
  private readonly _destroyRef = inject(DestroyRef);

  private _gridApi?: GridApi;

  readonly colDefs: ColDef<IRow>[] = [
    { field: 'id' },
    { field: 'name' },
    { field: 'email' },
  ];

  onGridReady(event: GridReadyEvent): void {
    this._gridApi = event.api;

    this._loadData();
  }

  private _loadData(): void {
    this._gridApi?.showLoadingOverlay();

    this._httpClient
      .get<IRow[]>('https://jsonplaceholder.typicode.com/users')
      .pipe(
        take(1),
        catchError(() => of([])),
        takeUntilDestroyed(this._destroyRef)
      )
      .subscribe((data) => {
        this._gridApi?.setGridOption('rowData', data);
      });
  }
}

```

app-root.component.spec.ts

```typescript
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AppRootComponent } from './app-root.component';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';

describe('AppRootComponent', () => {
  let component: AppRootComponent;
  let fixture: ComponentFixture<AppRootComponent>;
  let httpClient: HttpClient;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [],
      imports: [AppRootComponent, HttpClientTestingModule],
    }).compileComponents();

    fixture = TestBed.createComponent(AppRootComponent);
    component = fixture.componentInstance;
    httpClient = TestBed.inject(HttpClient);
  });

  it('should create component', () => {
    expect(component).toBeTruthy();
    expect.assertions(1);
  });

  it('should load data correctly', async () => {
    const loadDataSpy = jest.spyOn(component, '_loadData' as any);
    jest
      .spyOn(httpClient, 'get')
      .mockReturnValue(
        of([{ id: 1, name: 'john doe', email: 'john@mail.com' }])
      );

    fixture.detectChanges();
    await fixture.whenStable();

    expect(loadDataSpy).toHaveBeenCalled();
    expect.assertions(1);
  });
});
```

and the comand to run in console is:

```shell
npm install && jest --passWithNoTests
```

**Comment by n19htz** [comment]
fixed [reproduction](https://stackblitz.com/edit/stackblitz-starters-wjsbqj?file=src%2Fapp-root.componet.spec.ts)

**Comment by StephenCooper** [comment]
Thanks for the reproduction. I am investigating.

**Comment by StephenCooper** [comment]
I understand why the changes in v31.02 have caused this. 

Now that AG Grid internally runs outside of ZoneJs the setup of the grid no longer fills the ZoneJs queue with events. This means that when the test comes to ` await fixture.whenStable()` there is nothing in the ZoneJs queue and so the test continues.

This basically makes the test synchronous and the grid is destroyed before the `onGridReady` event can be fired. When the `AgGridAngular` component is destroyed it swallows events. I can see that this swallowing is preventing the events being fired. 

However, the main issue is that there is nothing within ZoneJs to hold up the execution of the test.  A unsatisfactory workaround to demonstrate this is as follows. 

```
  // setup a timer that will sit in ZoneJs and prevent it from being stable
  timer = setTimeout(() => {}, 100000);

  onGridReady(event: GridReadyEvent): void {
    this._gridApi = event.api;
    this._loadData();

    // Now that grid ready has fired clear the timer allowing fixture.whenStable() to pass
    clearTimeout(this.timer);
  }
```

https://stackblitz.com/edit/stackblitz-starters-kmpcjv?file=src%2Fapp-root.component.ts

I am going to continue looking into this to find other options.

**Comment by Horsty80** [comment]
> I understand why the changes in v31.02 have caused this.
> 
> Now that AG Grid internally runs outside of ZoneJs the setup of the grid no longer fills the ZoneJs queue with events. This means that when the test comes to ` await fixture.whenStable()` there is nothing in the ZoneJs queue and so the test continues.
> 
> This basically makes the test synchronous and the grid is destroyed before the `onGridReady` event can be fired. When the `AgGridAngular` component is destroyed it swallows events. I can see that this swallowing is preventing the events being fired.
> 
> However, the main issue is that there is nothing within ZoneJs to hold up the execution of the test. A unsatisfactory workaround to demonstrate this is as follows.
> 
> ```
>   // setup a timer that will sit in ZoneJs and prevent it from being stable
>   timer = setTimeout(() => {}, 100000);
> 
>   onGridReady(event: GridReadyEvent): void {
>     this._gridApi = event.api;
>     this._loadData();
> 
>     // Now that grid ready has fired clear the timer allowing fixture.whenStable() to pass
>     clearTimeout(this.timer);
>   }
> ```
> 
> https://stackblitz.com/edit/stackblitz-starters-kmpcjv?file=src%2Fapp-root.component.ts
> 
> I am going to continue looking into this to find other options.

Hello, i face similar issue that meens, when update to v31.0.2 all my test are breaking
I'm using React, but basically the issue is about the render of the grid

Basicaly a simple test that show i've no rows to show is KO

```js
describe('aggrid', () => {
  test('Should render no rows to show when no data', async () => {
    render(
      <MyCustomWrapperOfAggrid
        deals={[]} // is the rowData inside aggrid
      />,
    );

    await screen.findByText('No Rows To Show');
    expect(screen.queryByText('No Rows To Show')).toBeInTheDocument();
    const element = document.querySelector('.ag-overlay-no-rows-center')
    console.log('prettyDOM(element));
  });
```

If i log the same target component before and after the update i get this dom element

```html
// Before
<span
  class="ag-overlay-no-rows-center"
>
  No Rows To Show
</span>

// After
<span class="ag-overlay-no-rows-center" />

```

If this comment can help you to investigate :) 

let me see if you need other info

**Comment by StephenCooper** [comment]
@Horsty80 I believe this is a separate issue and would encourage this to be raised as a separate ticket. As you say you are using React so you will not be hitting the same issue which lives in the Angular part of AG Grid.

**Comment by n19htz** [comment]
> I understand why the changes in v31.02 have caused this.
> 
> Now that AG Grid internally runs outside of ZoneJs the setup of the grid no longer fills the ZoneJs queue with events. This means that when the test comes to ` await fixture.whenStable()` there is nothing in the ZoneJs queue and so the test continues.
> 
> This basically makes the test synchronous and the grid is destroyed before the `onGridReady` event can be fired. When the `AgGridAngular` component is destroyed it swallows events. I can see that this swallowing is preventing the events being fired.
> 
> However, the main issue is that there is nothing within ZoneJs to hold up the execution of the test. A unsatisfactory workaround to demonstrate this is as follows.
> 
> ```
>   // setup a timer that will sit in ZoneJs and prevent it from being stable
>   timer = setTimeout(() => {}, 100000);
> 
>   onGridReady(event: GridReadyEvent): void {
>     this._gridApi = event.api;
>     this._loadData();
> 
>     // Now that grid ready has fired clear the timer allowing fixture.whenStable() to pass
>     clearTimeout(this.timer);
>   }
> ```
> 
> https://stackblitz.com/edit/stackblitz-starters-kmpcjv?file=src%2Fapp-root.component.ts
> 
> I am going to continue looking into this to find other options.

this makes impossible to test properly data displayed by ag-grid entirely

**Comment by StephenCooper** [comment]
Yes, this breaks the current pattern so we will come back with a solution. 

Out of interest what are you aiming to validate with this test? That the onGridReady method is correctly hooked into the grid? Do you have other tests that also rely on when stable that don't involve onGridReady?

We will be revisiting the testing section of our docs and will likely encourage the use of e2e tests that work against the DOM instead of interacting with components directly. 

We have found many gotchas with these low level tests, especially around async handling. 

However, just to reiterate we will provide a solution to resolve this issue and get the tests working again.

**Comment by n19htz** [comment]
> Out of interest what are you aiming to validate with this test? That the onGridReady method is correctly hooked into the grid?
> 
> We will be revisiting the testing section of our docs and will likely encourage the use of e2e tests that work against the DOM instead of interacting with components directly.

Its intergation test, just for the minimal reproduction I used assertion involving spy on the component method. In the real world you query `DOM` for the "real" values displayed on the table. But to be able to see values on the table - `gridReady` event must be triggered, so the actual data is loaded and set to tables `rowData`. Most of our tables loading data from the server at `gridReady` event and to make meaningful assertions you need table with data. `For example`: We got table, at the `gridReady `event it loads users from the server and puts data into the table (like in reproduction example). After you select user row you can press Delete button, which will open confirm dialog and upon confirming a new http request is made. After request finished successfully it will apply delete transaction. 

So the example of test case here can be like this:

`it('should correctly delete user')`

1) assert that table displays 10 rows of users
2) trigger selection event on ag-grid for certain user row
3) assert that row selected and Delete button is not disabled
4) trigger click event on Delete button
5) assert that confirm dialog opened
6) trigger click event on Confirm button
7) assert that request with right payload is made
8) assert that table now displays 9 rows

we got **A LOT** of tests bound to that `gridReady` event

> Do you have other tests that also rely on when stable that don't involve onGridReady?

yes

**Comment by StephenCooper** [comment]
@n19htz I have made some changes and I wondered if you would like to test them against your project?

To use this file save it locally and update your package.json to have

```json
"ag-grid-angular": "file:{PATH_TO_FOLDER}/ag-grid-angular-31.0.3_rc.tgz",
```

These changes work against the reproduction provided. (Apart from the snapshot but I believe that is a separate issue to what you have reported here.)

If further tests fail with these changes that would be useful to know before we release a patch. 

[ag-grid-angular-31.0.3_rc.tgz](https://github.com/ag-grid/ag-grid/files/14013198/ag-grid-angular-31.0.3_rc.tgz)

**Comment by n19htz** [comment]
> @n19htz I have made some changes and I wondered if you would like to test them against your project?
> 
> To use this file save it locally and update your package.json to have
> 
> ```json
> "ag-grid-angular": "file:{PATH_TO_FOLDER}/ag-grid-angular-31.0.3_rc.tgz",
> ```
> 
> These changes work against the reproduction provided. (Apart from the snapshot but I believe that is a separate issue to what you have reported here.)
> 
> If further tests fail with these changes that would be useful to know before we release a patch.
> 
> [ag-grid-angular-31.0.3_rc.tgz](https://github.com/ag-grid/ag-grid/files/14013198/ag-grid-angular-31.0.3_rc.tgz)

<img width="644" alt="Screenshot 2024-01-22 at 20 09 07" src="https://github.com/ag-grid/ag-grid/assets/39006390/a38f4e32-6746-495e-a15f-e8994eda6893">

we using ag-grid modules, importing only what is needed for differebt our modules, not sure how to deal with file you provided

**Comment by StephenCooper** [comment]
Let me provide you with the module version to use instead then. 

[ag-grid-community-angular-31.0.3_rc.tgz](https://github.com/ag-grid/ag-grid/files/14013327/ag-grid-community-angular-31.0.3_rc.tgz)

**Comment by n19htz** [comment]
> Let me provide you with the module version to use instead then.
> 
> [ag-grid-community-angular-31.0.3_rc.tgz](https://github.com/ag-grid/ag-grid/files/14013327/ag-grid-community-angular-31.0.3_rc.tgz)

do i need to update other modules to 31.0.2?

**Comment by StephenCooper** [comment]
You don't have to, but if you are running into build issues then that might resolve them.

**Comment by n19htz** [comment]
ok, ill try tommorrow and return with feedback

**Comment by n19htz** [comment]
@StephenCooper `gridReady` is now triggered, and most of the tests are working. However, the tests for tables with custom cell renderers have stopped working because they are basically not rendered (their `agInit` method is not called at all). As a result, the assertions that target elements of custom cell renderers are failing.

**Comment by StephenCooper** [comment]
So this looks like a different issue due to the column not being visible and so then not being rendered. I imagine if you set `suppressColumnVirtualisation=true` then that test will pass. 

What I am trying to work out is why this has now changed between 31.0.1 and 31.02 and if it is at all related to the Angular issue of some other change.

Thank you for the reproduction!

**Comment by n19htz** [comment]
> So this looks like a different issue due to the column not being visible and so then not being rendered. I imagine if you set `suppressColumnVirtualisation=true` then that test will pass.

Yes, it works when I set `suppressColumnVirtualisation=true`, but this isn't a viable solution for us due to tables with numerous columns.

**Comment by StephenCooper** [comment]
We have just released 31.0.3 which should resolve these issues.

**Comment by n19htz** [comment]
> We have just released 31.0.3 which should resolve these issues.

All tests passed. Thanks!

**Comment by StephenCooper** [comment]
Thank you for your help in getting this resolved with your reproductions!
